lsp.go

```diff
@@ -144,6 +148,144 @@ func (s *Server) handleInitialize(
+	reply jsonrpc2.Replier,
+	r jsonrpc2.Request,
+	log zerolog.Logger,
+) error {
+	var params lsp.DidOpenTextDocumentParams
+	if err := json.Unmarshal(r.Params(), &params); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal params")
+		return err
+	}
+
+	fname := params.TextDocument.URI.Filename()
+	content := params.TextDocument.Text
+
+	err := checkFile(fname, content)
+	return s.sendDiagnostics(ctx, params.TextDocument.URI, err)
+}
+
+func (s *Server) handleDocumentChange(
+	ctx context.Context,
+	reply jsonrpc2.Replier,
+	r jsonrpc2.Request,
+	log zerolog.Logger,
+) error {
+	var params lsp.DidChangeTextDocumentParams
+	if err := json.Unmarshal(r.Params(), &params); err != nil {
+		log.Error().Err(err).Msg("failed to unmarshal params")
+		return err
+	}
+
+	if len(params.ContentChanges) != 1 {
+		err := fmt.Errorf("unexpected content change length: %d", len(params.ContentChanges))
+		log.Error().Err(err).Send()
+		return err
+	}
+
+	content := params.ContentChanges[0].Text
+	fname := params.TextDocument.URI.Filename()
+
+	err := checkFile(fname, content)
+	return s.sendDiagnostics(ctx, params.TextDocument.URI, err)
+}
+
+func (s *Server) sendDiagnostics(ctx context.Context, uri lsp.URI, err error) error {
+	diags := []lsp.Diagnostic{}
+
+	if err != nil {
+		e, ok := err.(*errors.Error)
+		if ok {
+			log.Debug().Str("error", e.Detailed()).Msg("failed to parse hcl directory")
+
+			fileRange := lsp.Range{}
+			fileRange.Start.Line = uint32(e.FileRange.Start.Line) - 1
+			fileRange.Start.Character = uint32(e.FileRange.Start.Column) - 1
+			fileRange.End.Line = uint32(e.FileRange.End.Line) - 1
+			fileRange.End.Character = uint32(e.FileRange.End.Column) - 1
+
+			diags = append(diags, lsp.Diagnostic{
+				Message:  err.Error(),
+				Range:    fileRange,
+				Severity: lsp.DiagnosticSeverityError,
+				Source:   "linter",
+			})
+		} else {
+			// if err doesn't have a range we assume the file is ok.
+			// later we can send a proper info dialog showing the internal error.
+			log.Debug().Err(err).Msg("ignoring error")
+		}
+	}
+
+	err = s.conn.Notify(ctx, lsp.MethodTextDocumentPublishDiagnostics, lsp.PublishDiagnosticsParams{
+		URI:         uri,
+		Diagnostics: diags,
+	})
+
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to send diagnostics to the client.")
+	}
 
 	return nil
 }
+
+// checkFile checks if the given changed file has any errors.
+// It parses all files in the directory but the provided one is added manually
+// because it can be unsaved.
+func checkFile(fname string, content string) error {
+	dir := filepath.Dir(fname)
+	parser := hcl.NewTerramateParser(dir)
+	err := parser.AddFile(fname, []byte(content))
+	if err != nil {
+		log.Error().Err(err).Send()
+		return err
+	}
+
+	dirEntries, err := os.ReadDir(dir)
+	if err != nil {
+		log.Error().Msgf("adding directory to terramate parser: %s", err)
+		return err
+	}
+
+	log.Trace().Msg("looking for Terramate files")
+
+	for _, dirEntry := range dirEntries {
```

Comment:
This is very similar to the logic done [here](https://github.com/mineiros-io/terramate/blob/52981a5477285507e417370a8ca37e46ed4c3707/hcl/hcl.go#L653), maybe we could extract the listing dir + getting all terramate files logic into a separate function, and then that is used here + also there (thinking on  suffix changes spreading here and causing surprising behavior, but not a huge deal).

I imagine you are aware and is just rushing, so feel free to disregard.